packages/frontend/component/src/ui/editable/inline-edit.tsx

```diff
@@ -0,0 +1,226 @@
+  useRef,
+  useState,
+} from 'react';
+
+import Input from '../input';
+import * as styles from './inline-edit.css';
+
+export interface InlineEditHandle {
+  triggerEdit: () => void;
+}
+
+export interface InlineEditProps
+  extends Omit<HTMLAttributes<HTMLSpanElement>, 'onChange' | 'onInput'> {
+  /**
+   * Content to be displayed
+   */
+  value?: string;
+  /**
+   * Whether the content is editable
+   */
+  editable?: boolean;
+
+  onInput?: (v: string) => void;
+  onChange?: (v: string) => void;
+
+  /**
+   * Trigger edit by `click` or `doubleClick`
+   * @default `'doubleClick'`
+   */
+  trigger?: 'click' | 'doubleClick';
+
+  /**
+   * whether to auto select all text when trigger edit
+   */
+  autoSelect?: boolean;
+
+  /**
+   * Placeholder when value is empty
+   */
+  placeholder?: string;
+  /**
+   * Custom placeholder `className`
+   */
+  placeholderClassName?: string;
+  /**
+   * Custom placeholder `style`
+   */
+  placeholderStyle?: CSSProperties;
+
+  handleRef?: ForwardedRef<InlineEditHandle>;
+
+  /**
+   * Customize attrs for the input
+   */
+  inputAttrs?: Omit<HTMLAttributes<HTMLInputElement>, 'onChange' | 'onBlur'>;
+}
+
+export const InlineEdit = ({
+  value,
+  editable,
+  className,
+  style,
+  trigger = 'doubleClick',
+  autoSelect,
+
+  onInput,
+  onChange,
+
+  placeholder,
+  placeholderClassName,
+  placeholderStyle,
+
+  handleRef,
+  inputAttrs,
+
+  ...attrs
+}: InlineEditProps) => {
+  const [editing, setEditing] = useState(false);
+  const [editingValue, setEditingValue] = useState(value);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useImperativeHandle<InlineEditHandle, InlineEditHandle>(handleRef, () => ({
+    triggerEdit,
+  }));
+
+  const triggerEdit = useCallback(() => {
+    if (!editable) return;
+    setEditing(true);
+    setTimeout(() => {
+      inputRef.current?.focus();
+      autoSelect && inputRef.current?.select();
+    }, 0);
+  }, [autoSelect, editable]);
+
+  const onDoubleClick = useCallback(() => {
+    if (trigger !== 'doubleClick') return;
+    triggerEdit();
+  }, [triggerEdit, trigger]);
+
+  const onClick = useCallback(() => {
+    if (trigger !== 'click') return;
+    triggerEdit();
+  }, [triggerEdit, trigger]);
+
+  const submit = useCallback(() => {
+    onChange?.(editingValue || '');
+  }, [editingValue, onChange]);
+
+  const onEnter = useCallback(() => {
+    inputRef.current?.blur();
+  }, []);
+
+  const onBlur = useCallback(() => {
+    setEditing(false);
+    submit();
+    // to reset input's scroll position to match actual display
+    inputRef.current?.scrollTo(0, 0);
+  }, [submit]);
+
+  const inputHandler = useCallback(
+    (v: string) => {
+      setEditingValue(v);
+      onInput?.(v);
+    },
+    [onInput]
+  );
+
+  // update editing value when value prop changes
+  useEffect(() => {
+    setEditingValue(value);
+  }, [value]);
+
+  // to make sure input's style is the same as displayed text
+  const inputWrapperInheritsStyles = {
```

Comment:
using inline style to have higher precedence?

Reply:
yes, to override `<Input />`'s style and inherit displayed text.